runtime/obj.c

```diff
@@ -106,7 +106,7 @@ CAMLprim value caml_obj_with_tag(value new_tag_v, value arg)
 
   sz = Wosize_val(arg);
   tg = (tag_t)Long_val(new_tag_v);
-  if (sz == 0) CAMLreturn (arg);
+  if (sz == 0) CAMLreturn (Atom(tg));
```

Comment:
(I wondered whether this change would degrade performance sensibly; `Atom(tg)` extracts the value from a precomputed table, so I don't think we would notice the difference, and it's not clear that we use `set_tag` on empty blocks that often.)

Reply:
`Atom(tg)` doesn't actually touch the table: it returns the address of the table plus `tg` (0-field objects are pointers into the atom table). So I can't imagine this making any difference.